mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala

```diff
@@ -286,40 +288,120 @@ object MatrixFactorizationModel extends Loader[MatrixFactorizationModel] {
       srcFeatures: RDD[(Int, Array[Double])],
       dstFeatures: RDD[(Int, Array[Double])],
       num: Int): RDD[(Int, Array[(Int, Double)])] = {
-    val srcBlocks = blockify(srcFeatures)
-    val dstBlocks = blockify(dstFeatures)
-    val ratings = srcBlocks.cartesian(dstBlocks).flatMap { case (srcIter, dstIter) =>
-      val m = srcIter.size
-      val n = math.min(dstIter.size, num)
-      val output = new Array[(Int, (Int, Double))](m * n)
+    val srcBlocks = blockify(rank, srcFeatures).zipWithIndex()
+    val dstBlocks = blockify(rank, dstFeatures)
+    val ratings = srcBlocks.cartesian(dstBlocks).map {
+      case (((srcIds, srcFactors), index), (dstIds, dstFactors)) =>
+        val m = srcIds.length
+        val n = dstIds.length
+        val dstIdMatrix = new Array[Int](m * num)
+        val scoreMatrix = Array.fill[Double](m * num)(Double.NegativeInfinity)
+        val pq = new BoundedPriorityQueue[(Int, Double)](num)(Ordering.by(_._2))
+
+        val ratings = srcFactors.transpose.multiply(dstFactors)
+        var i = 0
+        var j = 0
+        while (i < m) {
+          var k = 0
+          while (k < n) {
+            pq += dstIds(k) -> ratings(i, k)
+            k += 1
+          }
+          var size = pq.size
+          while (size > 0) {
+            size -= 1
+            val factor = pq.poll()
```

Comment:
Hi @MLnick , thanks for your review.
My original test for sorted is using: pq.toArray.sorted(Ordering.By[(Int, Double), Double](-_._2)), 
because pq.toArray.sorted(-_._2) build error. Maybe there is boxing/unboxing, the  performance is very bad.
Now, I use pq.toArray.sortBy(-_._2), the performance is good than poll. this 25s vs poll 26s.
Thanks.